src/backend/distributed/transaction/transaction_management.c

```diff
@@ -730,16 +730,25 @@ static void
 PushSubXact(SubTransactionId subId)
 {
 	/*
-	 * We need to allocate these in TopTransactionContext instead of current
-	 * subxact's memory context. This is because AtSubCommit_Memory won't
-	 * delete the subxact's memory context unless it is empty, and this
+	 * We need to allocate these in a long-living memory context instead of
+	 * current subxact's memory context. This is because AtSubCommit_Memory
+	 * won't delete the subxact's memory context unless it is empty, and this
 	 * can cause in memory leaks. For emptiness it just checks if the memory
 	 * has been reset, and we cannot reset the subxact context since other
 	 * data can be in the context that are needed by upper commits.
 	 *
 	 * See https://github.com/citusdata/citus/issues/3999
+	 *
+	 * Also, any memory error that we might encounter here might leave
+	 * activeSubXactContexts in an inconsistent state and this might result in
```

Comment:
As far as I can tell, we're only preventing memory leaks, not changing anything about what's in activeSubXactContexts.

Reply:
It seems `AllocSetContextCreateExtended` does a `malloc` for initSize. Wouldn't it mean we can prevent OOMs might happen below via `palloc`?

Reply:
hmz, but if that's the issue we're solving shouldn't we be more worried about things like CoordinatedRemoteTransactionsSavepointRollback ?

The issue described by the comment suggests we're more concerned with OOM leaving the global variables in an inconsistent state.

Reply:
>  but if that's the issue we're solving

Tbh, we cannot reliably produce the issue we are trying to solve. So, we don't have a very specific thing to fix. We sometimes get back-traces, with workloads that involve OOMs, and lots of plpgsql functions with multiple levels of `exception` blocks. And, the `bt`s points to (sub)transaction callbacks.

I think one of our suspicion is that an OOM on these functions, could easily cause crashes in different ways. Sometimes, we get PANIC because we might get 5 OOMs/errors in these functions and Postgres does it somewhere [here](https://github.com/postgres/postgres/blob/b40baa96a7ad789718dcf59b1dd73bae15c3a7e1/src/backend/utils/error/elog.c#L433-L442). Sometimes NULL pointer access, or seg faults, which I suspect because of an OOM during the callbacks, the stack is not properly maintained. 

>  shouldn't we be more worried about things like CoordinatedRemoteTransactionsSavepointRollback ?

I agree that the fix should be covering all the sub-transaction functions if we are really worried about OOMs

Reply:
> shouldn't we be more worried about things like CoordinatedRemoteTransactionsSavepointRollback ?

I think we should be careful about the memory context being used not during "about" but "begin", so that accessing to a NULL pointer --saved at the beginning of the transaction-- doesn't cause a segfault.

And for the abort handlers, I think getting an OOM error is less problematic than the "begin", since it would cause a panic/fatal instead of a segfault in that case.

Reply:
PANIC as worse as a segfault

Reply:
If we allocate the state kept for the subxact in the overarching transaciont, could we then remove the pfree's from the PopSubXact callback?

Assuming the change of memory context we allocate the state in we shouldn't be as afraid anymore for leaks do we?

Changing the context we allocate in does not prevent an OOM from happening. This context could also be out of allocated space and failing to expand.

This is ofcourse assuming that an allocation failure causes a segfault, presumably due to double pfree.

Reply:
> Changing the context we allocate in does not prevent an OOM from happening. This context could also be out of allocated space and failing to expand.

Yes, of course that is possible, but with a lot less likelihood. I was also planning to calculate how much memory a single `SAVEPOINT/ROLLBACK` generates, so that we could consider a saner value for the initial allocation, which is 8KB for now.

> If we allocate the state kept for the subxact in the overarching transaciont, could we then remove the pfree's from the PopSubXact callback?
 Assuming the change of memory context we allocate the state in we shouldn't be as afraid anymore for leaks do we?

Sounds interesting, and probably right. I mean, the context we use is not relevant, even for `TopTransactionContext`, we could consider not `pfree`ing anyway. 


ps: I have not looked into the fix (and this code-path for a while), so I still suspect `pfree` is not the only root of these crashes.

Reply:
> so I still suspect pfree is not the only root of these crashes.

This is true, linitial() might also trigger a segfault if the stack is empty (as given in the stack trace at PR description)

Reply:
> Yes, of course that is possible, but with a lot less likelihood. I was also planning to calculate how much memory a single SAVEPOINT/ROLLBACK generates, so that we could consider a saner value for the initial allocation, which is 8KB for now.

presumably we know the maximum length that these commands can have, so we should be able to allocate them on the stack

Reply:
> This is ofcourse assuming that an allocation failure causes a segfault

But this is true even in the double-free case, since the main problem there is that we leave PushSubXact() before setting all the globar vars properly, and this could only happen due to an OOM error ?